src/context/data-context.jsx

```diff
@@ -0,0 +1,75 @@
+import { createContext, useContext, useEffect, useReducer } from "react";
+import { db } from "../config/firebase-config";
+import { collection, onSnapshot } from "firebase/firestore";
+import { useAuth } from "./auth-context";
+import { dataReducer, initialReducerData } from "../reducer/dataReducer";
+
+const DataContext = createContext({
+  state: initialReducerData,
+  dispatch: () => {},
+});
+
+const DataProvider = ({ children }) => {
+  const { user } = useAuth();
+
+  const [state, dispatch] = useReducer(dataReducer, initialReducerData);
+
+  if (user) {
+    onSnapshot(
+      collection(db, "users", `${user.uid}`, "savedNotes"),
+      (snapshot) => {
+        dispatch({
+          type: "INITIALIZE_SAVED_NOTES",
+          payload: snapshot.docs.map((doc) => ({ ...doc.data(), id: doc.id })),
+        });
+      }
+    );
+
+    onSnapshot(
+      collection(db, "users", `${user.uid}`, "archivedNotes"),
+      (snapshot) => {
+        dispatch({
+          type: "INITIALIZE_ARCHIVED_NOTES",
+          payload: snapshot.docs.map((doc) => ({ ...doc.data(), id: doc.id })),
+        });
+      }
+    );
+
+    onSnapshot(
+      collection(db, "users", `${user.uid}`, "trashedNotes"),
+      (snapshot) => {
+        dispatch({
+          type: "INITIALIZE_TRASHED_NOTES",
+          payload: snapshot.docs.map((doc) => ({ ...doc.data(), id: doc.id })),
+        });
+      }
+    );
+
+    onSnapshot(collection(db, "users", `${user.uid}`, "tags"), (snapshot) => {
+      const list = snapshot.docs.map((doc) => doc.data());
+      dispatch({
+        type: "INITIALIZE_TAGS_LIST",
+        payload: list[0].tagsList,
+      });
+    });
+  }
```

Comment:
Uncaught TypeError: Cannot read properties of undefined (reading 'tagsList')
    at Array.<anonymous> (main.43e34f35.js:2:948322)
    at next (main.43e34f35.js:2:819677)
    at main.43e34f35.js:2:785302

Getting this error in the console. Just check this out.

Reply:
There is 2 seconds delay in getting data from firebase. That's why it gives errors the first time around. 
But fixed it. Thanks for pointing it out